File: graphs/number_of_islands.rb
# @param {Character[][]} grid
# @return {Integer}
def dfs(grid, i, j)
    n = grid.size
    m = grid[0].size

    if i < 0 || j < 0 || i >= n || j >= m || grid[i][j] != '1'
        return
    end

    grid[i][j] = '2' # mark visited

    # explore neighbors
    dfs(grid, i + 1, j) # bottom
    dfs(grid, i - 1, j) # top
    dfs(grid, i, j + 1) # right
    dfs(grid, i, j - 1) # left
end

def num_islands(grid)
    count = 0
    n = grid.size
    m = grid[0].size

    for i in 0...n
        for j in 0...m
            if grid[i][j] == '1'
                count += 1
                dfs(grid, i, j)
            end
        end
    end
    return count
end

File: graphs/number_of_islands.py
class Solution:
    def dfs(self, grid, i, j):
        n = len(grid)
        m = len(grid[0])

        if i < 0 or j < 0 or i >= n or j >= m or grid[i][j] != '1':
            return

        grid[i][j] = '2'  # mark visited

        # explore neighbors
        self.dfs(grid, i + 1, j)  # bottom
        self.dfs(grid, i - 1, j)  # top
        self.dfs(grid, i, j + 1)  # right
        self.dfs(grid, i, j - 1)  # left

    def numIslands(self, grid: List[List[str]]) -> int:
        if not grid or not grid[0]:
            return 0

        count = 0
        n = len(grid)
        m = len(grid[0])

        for i in range(n):
            for j in range(m):
                if grid[i][j] == '1':
                    count += 1
                    self.dfs(grid, i, j)
        return count